README.md
# NAME

Monoceros - PSGI/Plack server with event driven connection manager, preforking workers

# SYNOPSIS

    % plackup -s Monoceros --max-keepalive-reqs=10000 --max-workers=2 -a app.psgi

# DESCRIPTION

Monoceros is PSGI/Plack server supports HTTP/1.1. Monoceros has a event-driven 
connection manager and preforking workers. Monoceros can keep large amount of 
connection at minimal processes.

                                                          +--------+
                                                      +---+ worker |
          TCP       +---------+   UNIX DOMAIN SOCKET  |   +--------+
    --------------- | manager | ----------------------+ 
                    +---------+                       |   +--------+
    <- keepalive ->              <-- passing fds -->  `---+ worker |
                                                          +--------+

Features of Monoceros

\- a manager process based on [AnyEvent](https://metacpan.org/pod/AnyEvent) keeps over C10K connections

\- uses [IO::FDPass](https://metacpan.org/pod/IO%3A%3AFDPass) for passing a file descriptor to workers

\- supports HTTP/1.1 and also supports HTTP/1.0 keepalive

And this server inherit [Starlet](https://metacpan.org/pod/Starlet). Monoceros supports following features too.

\- prefork and graceful shutdown using [Parallel::Prefork](https://metacpan.org/pod/Parallel%3A%3APrefork)

\- hot deploy using [Server::Starter](https://metacpan.org/pod/Server%3A%3AStarter)

\- fast HTTP processing using [HTTP::Parser::XS](https://metacpan.org/pod/HTTP%3A%3AParser%3A%3AXS) (optional)

\- accept4(2) using [Linux::Socket::Accept4](https://metacpan.org/pod/Linux%3A%3ASocket%3A%3AAccept4) (optional)

Currently, Monoceros does not support spawn-interval and max-keepalive-reqs.

# COMMAND LINE OPTIONS

In addition to the options supported by [plackup](https://metacpan.org/pod/plackup), Monoceros accepts following options(s).
Note, the default value of several options is different from Starlet.

## --max-workers=#

number of worker processes (default: 5)

## --timeout=#

seconds until timeout (default: 300)

## --keepalive-timeout=#

timeout for persistent connections (default: 10)

## --max-reqs-per-child=#

max. number of requests to be handled before a worker process exits (default: 1000)

## --min-reqs-per-child=#

if set, randomizes the number of requests handled by a single worker process between the value and that supplied by `--max-reqs-per-chlid` (default: none)

## --max-keepalive-connection=#

max, number of connections to keep in the manager process. If you want to increase this value, You should check your system limitations. (default: half number of POSIX::\_SC\_OPEN\_MAX)

## --max-readahead-reqs=#

max. number of requests to continue to read a request in a worker process. Monoceros can read a next request after the response for maximum throughput. (default: 100)

## --min-readahead-reqs=#

if set, randomizes the number of requests to continue to read a request between the value and that supplied by `--max-readahead-reqs` (default: none)

## --disable-keepalive

disable HTTP keepalive feature (default: enabled)

# RECOMMENDED MODULES

For more performance. I recommends you to install these module.

\- [EV](https://metacpan.org/pod/EV)

\- [HTTP::Parser::XS](https://metacpan.org/pod/HTTP%3A%3AParser%3A%3AXS)

# EXPERIMENTAL FEATURES

- accept4

    Linux only. If Linux::Socket::Accept4 is available, Monoceros use it.
    accept4(2) can do accept(2) and set SOCK\_CLOEXEC|SOCK\_NONBLOCK at once. 

- sendfile

    Monoceros send IO::Handle like body with sendfile(2). [Sys::Sendfile](https://metacpan.org/pod/Sys%3A%3ASendfile) is required.

# SEE ALSO

[Starlet](https://metacpan.org/pod/Starlet), [Server::Starter](https://metacpan.org/pod/Server%3A%3AStarter), [AnyEvent](https://metacpan.org/pod/AnyEvent), [IO::FDPass](https://metacpan.org/pod/IO%3A%3AFDPass)

# LICENSE      

Copyright (C) Masahiro Nagano

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Masahiro Nagano <kazeburo@gmail.com>


Changes
Revision history for Perl extension Monoceros

{{$NEXT}}

0.29 2023-11-09T09:55:36Z

  - Failsafes for stale process trees after reloads #9

0.28 2021-09-22T07:50:49Z

  - Fixed handling of max_requests_per_child and added issue specific tests (Thank you pepl)

0.27 2014-05-28T03:25:37Z

  - stop using POSIX::Socket

0.26 2013-10-23T01:33:36Z

  - add --disable-keepalive opt
  - change the default of --max-reqs-per-child, 100 to 1000
  - support sendfile(2) (EXPERIMENTAL)

0.25 2013-10-05T14:29:57Z

  - supports accept4(2) using Linux::Socket::Accept4

0.24 2013-09-02T06:39:06Z

  - supports harakiri

0.23 2013-08-29T04:45:27Z

  - no trial version

0.22 2013-08-29T02:50:57Z

  - improved around readahead

0.21 2013-08-28T05:21:59Z

  - use AnyEvent::Util::portable_socketpair

0.20 2013-08-28T02:40:00Z

  - refactor: remove unused variables
  - use sockpair instead of a internal unix-domain server
  - change stats file format

0.19 2013-06-11T17:18:23Z

  - use file stat instead of symlink
  - back fdopen in worker. error occurred in *bsd 

0.18 2013-06-11T06:12:22Z

  - support min-readahead-reqs option
  - requires Test::TCP >= 2.00

0.17 2013-06-07T01:50:12Z

  - support max-readahead-reqs option
  - change status content in MonocerosStatus
  - require POSIX::getpeername

0.16 2013-05-28T01:04:48Z

  - remove sockfile at process exit

0.15 2013-05-23T04:39:30Z

  - fix pod

0.14 2013-05-23T02:38:10Z

  - add Plack::Middleware::MonocerosStatus to display manager status

0.13 2013-05-22T06:16:38Z

  - add max-keepalive-connection option

0.12 2013-05-20T07:25:35Z

  - refactoring worker->manager socket passing

0.11 2013-05-17T05:37:59Z

  - close timeouted sockets in preforked workers.
    avoid nginx revc failed error message

0.10 2013-05-16T05:35:01Z

  - refactoring around read-forward. Improve performance in < 10 connections.

0.09 2013-05-15T06:16:35Z

  - support HTTP/1.1

0.08 2013-05-13T16:49:17Z

  - read-forward next reqs/socket before select
  - drop max-keepalive-reqs
  - bugfix around graceful-stop   

0.07 2013-05-10T11:25:39Z

  - refactoring around accepted sockets in worker.

0.06 2013-05-10T06:03:49Z

  - bugfix. recv() incorrect fd

0.05 2013-05-10T02:28:02Z

  - pre-sysread for tcp_defer_accept sockets.
    if sysread returns EAGAIN. push a socket to connection manager process.
    relate url: https://bugs.launchpad.net/ubuntu/+source/apache2/+bug/134274
                https://forums.aws.amazon.com/thread.jspa?threadID=45564

0.04 2013-05-09T04:28:31Z

   - refactoring around timeout and disconnected sockets

0.03 2013-05-09T01:46:46Z

   - accept(2) in workers if enabled defer_accept.

0.02 2013-05-07T07:35:30Z

   - refactoring & performance tuning 

0.01 2013-04-26T05:03:45Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Monoceros',
    module_name     => 'Monoceros',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Plack/Middleware/MonocerosStatus.pm
package Plack::Middleware::MonocerosStatus;

use strict;
use warnings;
use parent qw(Plack::Middleware);
use Plack::Util::Accessor qw(path allow);
use Net::CIDR::Lite;
use IO::Socket::UNIX;

sub prepare_app {
    my $self = shift;
    if ( $self->allow ) {
        my @ip = ref $self->allow ? @{$self->allow} : ($self->allow);
        my @ipv4;
        my @ipv6;
        for (@ip) {
            # hacky check, but actual checks are done in Net::CIDR::Lite.
            if (/:/) {
                push @ipv6, $_;
            } else {
                push @ipv4, $_;
            }
        }
        if ( @ipv4 ) {
            my $cidr4 = Net::CIDR::Lite->new();
            $cidr4->add_any($_) for @ipv4;
            $self->{__cidr4} = $cidr4;
        }
        if ( @ipv6 ) {
            my $cidr6 = Net::CIDR::Lite->new();
            $cidr6->add_any($_) for @ipv6;
            $self->{__cidr6} = $cidr6;
        }
    }
}

sub call {
    my ($self, $env) = @_;
    if( $self->path && $env->{PATH_INFO} eq $self->path ) {
        return $self->_handle_status($env);
    }
    $self->app->($env);
}

sub _handle_status {
    my ($self, $env ) = @_;

    if ( ! $self->allowed($env->{REMOTE_ADDR}) ) {
        return [403, ['Content-Type' => 'text/plain'], [ 'Forbidden' ]];
    }
    
    if ( !$env->{X_MONOCEROS_WORKER_STATS} || ! -f $env->{X_MONOCEROS_WORKER_STATS}) {
        return [500, ['Content-Type' => 'text/plain'], [ 'Monoceros stats file not found' ]];
    }

    open(my $fh, $env->{X_MONOCEROS_WORKER_STATS}) or 
        return [500, ['Content-Type' => 'text/plain'], [ 'Could not open Monoceros stats: $!' ]];
    my $len = $fh->sysread(my $buf, 1024);
    if ( !$len ) {
        return [500, ['Content-Type' => 'text/plain'], [ 'Could not read stats: $!' ]];
    }
    my %stats;
    for my $str ( split /&/, $buf ) {
        my ($key,$val) = split /=/, $str, 2;
        $stats{$key} = $val;
    }
    my $msg = "Total: ".$stats{total}."\015\012";
    $msg .= "Waiting: ".$stats{waiting}."\015\012";
    $msg .= "Processing: ".$stats{processing}."\015\012";
    $msg .= "MaxWorkers: ".$stats{max_workers}."\015\012\015\012";
    
    return [200, ['Content-Type' => 'text/plain'], [$msg]];
}

sub allowed {
    my ( $self , $address ) = @_;
    if ( $address =~ /:/) {
        return unless $self->{__cidr6};
        return $self->{__cidr6}->find( $address );
    }
    return unless $self->{__cidr4};
    return $self->{__cidr4}->find( $address );
}


1;

__END__

=head1 NAME

Plack::Middleware::MonocerosStatus - show Monoceros connection manager status

=head1 SYNOPSIS

  use Plack::Builder;

  builder {
      enable "MonocerosStatus",
          path => '/monoceros-status',
          allow => [ '127.0.0.1', '192.168.0.0/16' ],
      $app;
  };

  % curl http://server:port/monoceros-status
  Total: 100
  Waiting: 98
  Processing: 2

=head1 DESCRIPTION

Plack::Middleware::MonocerosStatus is a middleware to display Monoceros manager status.

  Waiting - a number of keepalive socket, waiting to next request
  Processing - a number of request arrived socket, sockets are queued or processed by worker
  Total - Waiting + Processing

=head1 CONFIGURATIONS

=over 4

=item path

  path => '/server-status',

location that displays server status

=item allow

  allow => '127.0.0.1'
  allow => ['192.168.0.0/16', '10.0.0.0/8']

host based access control of a page of server status. supports IPv6 address.

=back

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo {at} gmail.comE<gt>

=head1 SEE ALSO

L<Monoceros>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Monoceros.pm
package Monoceros;

use strict;
use warnings;
use 5.008005;

our $VERSION = "0.29";

1;
__END__

=encoding utf-8

=head1 NAME

Monoceros - PSGI/Plack server with event driven connection manager, preforking workers

=head1 SYNOPSIS

    % plackup -s Monoceros --max-keepalive-reqs=10000 --max-workers=2 -a app.psgi

=head1 DESCRIPTION

Monoceros is PSGI/Plack server supports HTTP/1.1. Monoceros has a event-driven 
connection manager and preforking workers. Monoceros can keep large amount of 
connection at minimal processes.

                                                          +--------+
                                                      +---+ worker |
          TCP       +---------+   UNIX DOMAIN SOCKET  |   +--------+
    --------------- | manager | ----------------------+ 
                    +---------+                       |   +--------+
    <- keepalive ->              <-- passing fds -->  `---+ worker |
                                                          +--------+

Features of Monoceros

- a manager process based on L<AnyEvent> keeps over C10K connections

- uses L<IO::FDPass> for passing a file descriptor to workers

- supports HTTP/1.1 and also supports HTTP/1.0 keepalive

And this server inherit L<Starlet>. Monoceros supports following features too.

- prefork and graceful shutdown using L<Parallel::Prefork>

- hot deploy using L<Server::Starter>

- fast HTTP processing using L<HTTP::Parser::XS> (optional)

- accept4(2) using L<Linux::Socket::Accept4> (optional)

Currently, Monoceros does not support spawn-interval and max-keepalive-reqs.

=head1 COMMAND LINE OPTIONS

In addition to the options supported by L<plackup>, Monoceros accepts following options(s).
Note, the default value of several options is different from Starlet.

=head2 --max-workers=#

number of worker processes (default: 5)

=head2 --timeout=#

seconds until timeout (default: 300)

=head2 --keepalive-timeout=#

timeout for persistent connections (default: 10)

=head2 --max-reqs-per-child=#

max. number of requests to be handled before a worker process exits (default: 1000)

=head2 --min-reqs-per-child=#

if set, randomizes the number of requests handled by a single worker process between the value and that supplied by C<--max-reqs-per-chlid> (default: none)

=head2 --max-keepalive-connection=#

max, number of connections to keep in the manager process. If you want to increase this value, You should check your system limitations. (default: half number of POSIX::_SC_OPEN_MAX)

=head2 --max-readahead-reqs=#

max. number of requests to continue to read a request in a worker process. Monoceros can read a next request after the response for maximum throughput. (default: 100)

=head2 --min-readahead-reqs=#

if set, randomizes the number of requests to continue to read a request between the value and that supplied by C<--max-readahead-reqs> (default: none)

=head2 --disable-keepalive

disable HTTP keepalive feature (default: enabled)

=head1 RECOMMENDED MODULES

For more performance. I recommends you to install these module.

- L<EV>

- L<HTTP::Parser::XS>

=head1 EXPERIMENTAL FEATURES

=over 4

=item accept4

Linux only. If Linux::Socket::Accept4 is available, Monoceros use it.
accept4(2) can do accept(2) and set SOCK_CLOEXEC|SOCK_NONBLOCK at once. 

=item sendfile

Monoceros send IO::Handle like body with sendfile(2). L<Sys::Sendfile> is required.

=back

=head1 SEE ALSO

L<Starlet>, L<Server::Starter>, L<AnyEvent>, L<IO::FDPass>

=head1 LICENSE      

Copyright (C) Masahiro Nagano

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

